### ISSUE TITLE
Remove book-all-books-block.tpl.php (replace with theme function)

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#546]
We have a single template file provided by book module, `book-all-books-block.tpl.php`.  This template provides only a single nav tag.  This is the kind of thing that is likely to upset people, making them dig through core to find the source of a single nav tag. This is a block, people can specify the tag they want in the blocks UI now, or by overriding a block template like `block--book--navigation.tpl.php`.

**Comment by quicksketch** [comment]
>  I'm not sure how many people even theme this block, but it would certainly be faster as a theme function. 

The benefit here is negligible on an element that's only going to be shown at most once per page. Aren't templates generally better for themers?

**Comment by jenlampton** [comment]
Not if they only contain a single nav tag :(

**Comment by quicksketch** [comment]
I don't think this was intended to be closed yet?

**Comment by jenlampton** [comment]
No, sorry, meant to hit "Comment".

My point was that - the benefit of using a template here is negligible, and the fact that this insignificant nav tag gets a template causes confusion. 

Templates should be used for things that are important - things people often theme, and that are easier to theme in template form than as a function.  If a theme dev looks in core for templates to override, they should find the most important things like node, comment, user, taxonomy-term, but certainly not book-all-books-block.

**Comment by quicksketch** [comment]
PR at https://github.com/backdrop/backdrop/pull/641 needs a reroll (doesn't merge currently).

Still not 100% sure about this, I'd think that more templates was always a good thing, but I'm ambivalent. Long-term, this entire block should probably be replaced with a view anyway.

**Comment by jenlampton** [comment]
In this case it's more templates adding more markp, all of which is unnecessary. If people want a nav tag, they can add one in the blocks UI, or by overriding `block--book--navigation.tpl.php`.  There's no reason for this extra layer of theme overrides.

New PR removes the nav tag entirely - but because we've now got a markup change going on, this may need to wait for 2.x?

**Comment by quicksketch** [comment]
> In this case it's more templates adding more markp, all of which is unnecessary. If people want a nav tag, they can add one in the blocks UI, or by overriding block--book--navigation.tpl.php. There's no reason for this extra layer of theme overrides.

Can you outline how it would be possible to add a `<nav>` tag around each book menu, like the one we removed, if you were a themer? Whenever we remove a template, there should be some way to override that output still. Usually through a theme suggestion or similar. I can't see how that would be (easily) possible after removing this template. I looks to me like to do that as a themer I would:

- Have to `hook_block_view_alter()` to add a `#theme`
- Add an entry to `hook_theme_info()`.
- Make a new template.

Basically, you'd have to redo everything that we've removed here.

**Comment by quicksketch** [comment]
> Basically, you'd have to redo everything that we've removed here.

I looked into this and there is *sort of* a way around it. You could override `theme_menu_tree()` in your template.php to add the missing `<nav>` tags. But there's no way to do it just for Book module's menus.

Here's the code in `menu.inc` that adds the call to `theme_menu_tree()`:

```
    // Add the theme wrapper for outer markup.
    // Allow menu-specific theme overrides.
    $build['#theme_wrappers'][] = 'menu_tree__' . strtr($data['link']['menu_name'], '-', '_');
```

We could replicate the exact same markup if:

- We added another suggestion here for `$data['link']['#original_link']['module']`.
- Then book module provided a `menu-tree--book.tpl.php` file.

However, I don't much like this because I find that working with a larger outer loop (like the current template) is easier to work with than an inner template.

**Comment by jenlampton** [comment]
I just hate having a template file that outputs a single HTML tag. it's sloppy.  Let's wait until we rethink books (outline?) and solve the problem then.
